lua/ttt2/libraries/bodysearch.lua

```diff
@@ -0,0 +1,1007 @@
+	-- @note This function is old and should be redone on a scoreboard rework
+	-- @realm client
+	function bodysearch.PreprocSearch(raw)
+		local search = {}
+
+		for i = 1, #bodysearch.searchResultOrder do
+			local type = bodysearch.searchResultOrder[i]
+			local searchData = bodysearch.GetContentFromData(type, raw)
+
+			if not searchData then continue end
+
+			-- a workaround to build the rext for the scoreboard
+			local text = searchData.text.text
+			local transText = ""
+
+			-- only use the first text entry here
+			local par = text[1].params
+			if par then
+				-- process params (translation)
+				for k, v in pairs(par) do
+					par[k] = LANG.TryTranslation(v)
+				end
+
+				transText = transText .. LANG.GetParamTranslation(text[1].body, par) .. " "
+			else
+				transText = transText .. LANG.TryTranslation(text[1].body) .. " "
+			end
+
+			search[type] = {
+				img = searchData.iconMaterial:GetName(),
+				text = transText,
+				p = i -- sorting number
+			}
+
+			-- special cases with icon text
+			local iconTextFn = TypeToIconText(type, raw)
+			if isfunction(iconTextFn) then
+				search[type].text_icon = iconTextFn()
+			end
+		end
+
+		---
+		-- @realm client
+		hook.Run("TTTBodySearchPopulate", search, raw)
+
+		return search
+	end
+
+	---
+	-- This function handles the storing of the streamed search result data. It takes into account
+	-- who searched the body and with which role. Also new data is appended to existing data to only
+	-- improve what a player knows, but to never remove information on.
+	-- This functions considers the roles and the settings of the local player and the player that
+	-- inspected the body.
+	-- @param table sData The table of scene data that should be stored
+	-- @note The data is stored as `bodySearchResult` on the ragdoll and the owner of the ragdoll
+	-- @realm client
+	function bodysearch.StoreSearchResult(sData)
+		if not sData.ragOwner then return end
+
+		-- if existing result was not ours, it was detective's, and should not
+		-- be overwritten
+		local ply = sData.ragOwner
+		local rag = sData.rag
+
+		-- do not store if searching player (client) is spectator
+		if LocalPlayer():IsSpec() then return end
+
+		-- if the currently stored search result is by a public policing role, it should be kept
+		-- it can be overwritten by another public policing role though
+		-- data can still be updated, but the original base is kept
+		local oldBase
+		if ply.bodySearchResult and ply.bodySearchResult.base and ply.bodySearchResult.base.isPublicPolicingSearch
+			and not sData.base.isPublicPolicingSearch
+		then
+			oldBase = sData.base
+		end
+
+		-- merge new data into old data
+		-- this is useful if a player had good data on a body from another source
+		-- and now gets updated info on it as it now only replaces the newly added
+		-- entries
+		local newData = ply.bodySearchResult or {}
+		table.Merge(newData, sData)
```

Comment:
the old way was to always overwirte it - but keep the old data when a detective previously searched it. This caused issues though in mode 2. Because previously abilable data was removed.

We either have to keep one of these two bugs, or we have to introduce a more complex solution to the problem

Reply:
Okay, but we would at least need to adjust this code section:
1. the base table is being modified by the table merge function, because it is invoked recursively (see https://github.com/Facepunch/garrysmod/blob/master/garrysmod/lua/includes/extensions/table.lua#L85-L99)
so you would need to e.g. copy the table above.
2. The name oldBase seems to be wrong here, because we are saving the possible new base data there. Rename this and the comments accordingly.